### DISCUSSION TITLE
Data extraction from medical reports

### DISCUSSION THREAD

**Discussion by farrandi** [explosion/spaCy#8635] [Help: Coding & Implementations] [answered]
Hey, 
I would like to extract data from medical reports. I have trained a NER model to help with this but I am not sure how to extract data with them.

I would like to get data in a format of `[Label] : [Value] [Units]` with the units being an optional entity and value ranging from numbers to words like: "mild","normal",  "severe", etc. (I want the [Label] and [Value Units] to be separate variables)

An example of some sentences with labeled entities are:
![image](https://user-images.githubusercontent.com/54917098/124838210-3c242b80-df3b-11eb-8ea8-50a9587db0ec.png)
Here I would like to extract: `Pulmonic valve structure` and  `Normal`  as relation (but 2 different variables) as well as `Pulmonic valve function` and `Normal` 

and a basic structure:
![image](https://user-images.githubusercontent.com/54917098/124838421-accb4800-df3b-11eb-887b-39f81526daa7.png)
Where extracting that would be `AV Peak Gradient` and `6.23 mmHg`

So I have tried several methods like using Matchers but they take the whole span, so I am not sure how to separate the label and the value. And the entities can have multiple tokens so I find it hard to match the token list and entity list together for extraction.

I also tried manually coding it for the last case and it worked, but I am not sure how to do the same for the sentence example.
```
def get_data(doc):
    all_data = []
    for index, entity in enumerate(doc.ents):
        if (entity.label_ == "OBS" or entity.label_ == "ANTMCL_STR") and index < len(doc.ents) - 2:
            ent_1 = doc.ents[index+1] 
            ent_2 = doc.ents[index+2]
            if (ent_1.label_ == "VALUE" and ent_2.label_ == "UNITS"):
                data = {
                    'ENTITY': str(doc.ents[index]),
                    'VALUE': str(doc.ents[index+1]),
                    'UNITS': str(doc.ents[index+2])
                }
                all_data.append(data)
    return all_data
```
Here, I just used the entity list and not tokens.
However, Matchers return start and end of tokens in the doc so I am not sure what to do. 

Any tips/ hints would be helpful. Thanks!

**Comment by polm** [comment, marked as answer]
Are the examples you give typical of complete documents, or sentences within longer documents?

Once you get tokens from a Matcher, you can go from a token to the containing doc or sentence, and from there get the entities. You can also get entities from spans. 

It looks like the main thing making your code complicated is that you're using fixed indexes, rather than accumulating entity types. You should make a dict (or `defaultdict(list)`) and add stuff from a relevant span to it. Something like this:

```
span = ... # you got this from the Matcher somehow
out = {} # this is where your observations go
for ent in span.ents:
    # first just skip ents if they're the wrong type
    if ent.label_ not in ("VALUE", ... whatever ... ): continue
    
    out[ent.label_] = ent.text

# at this point the out dict has all your info
```

Saving all the patterns in the dict without using token indices allows you to ignore word order, which makes the code dramatically simpler. It also makes multi-token entities a non-issue. 

For your given examples, it's clear that all the attributes are associated with each other, but that won't always be the case. For example you might have a sentence like:

> Patient took 30mg of Fixemup and 2 tubes of Painaway

And you need to associate "30mg" with "Fixemup", not with "Painaway". For that you should use the dependency parse; you can do that manually in code or use the dependency matcher. I suspect you'll see relatively few patterns, and working on the common ones first and seeing what coverage that gives you should give you a better idea of whether that's enough or whether you need to work on the problem more. (If you find the dependency parse confusing it's totally OK to use the normal matcher to match patterns like "AMOUNT FORMAT of DRUG", but the parse relations are quite powerful so I would recommend taking the time to learn them.)

Also definitely take a look at the [relation extraction example project](https://github.com/explosion/projects/tree/v3/tutorials/rel_component).
